feat(KONFLUX-3187): adding test selection logic for repo infra-deploy…

kasemAlem opened the pull request:
- Add Rules for the infra-deployments so changes in a specific component in infra-deployments runs it's suite
- If changes are out of components it runs the default konflux-demo suite 
- Fixed the getChangedFiles to look into infra-deployments adding to the e2e-tests

1- Examples from local Run : 
in the tmp/infra-deployments created a new branch , edited README of components/integration  , added committed 
ran the DryRun , so we expect that the **LabelFilter** have both "integration-service" and "konflux-demo": 

```
exec: git "-C" "/home/kasemalem/integration-project/e2e-tests/tmp/infra-deployments" "diff" "--name-status" "main..HEAD"
I0815 15:47:48.094713  520112 utils.go:474] The following files, components/integration/README.md, were changed!
I0815 15:47:48.094747  520112 types.go:72] Loading the catalog for, infra-deployments, from category, tests
I0815 15:47:48.094759  520112 types.go:144] The following rules have matched Infra-deployments PR Components File Diff Execution.

exec: ginkgo "--seed=1723726067" "--timeout=1h30m0s" "--grace-period=30s" "--output-interceptor-mode=none" "--label-filter=,integration-service,konflux-demo" "--junit-report=e2e-report.xml" "--p" "--output-dir=." "./cmd" "--"

```
 




2- Example of local Run :

Changing only the README.md file in Infra-deployments folder so we expect that only the **konflux-demo**  run 

```
exec: git "-C" "/home/kasemalem/integration-project/e2e-tests/tmp/infra-deployments" "diff" "--name-status" "main..HEAD"
I0815 16:01:32.736127  526636 utils.go:474] The following files, README.md, were changed!
I0815 16:01:32.736182  526636 types.go:72] Loading the catalog for, infra-deployments, from category, tests
I0815 16:01:32.736198  526636 types.go:144] The following rules have matched Infra Deployments Default Test Execution.
I0815 16:01:32.736203  526636 types.go:169] Will apply rules
exec: ginkgo "--seed=1723726891" "--timeout=1h30m0s" "--grace-period=30s" "--output-interceptor-mode=none" "--label-filter=konflux-demo" "--junit-report=e2e-report.xml" "--p" "--output-dir=." "./cmd" "--"
```


3- Example when no file was changed: 
```
exec: git "-C" "/home/kasemalem/integration-project/e2e-tests/tmp/infra-deployments" "diff" "--name-status" "main..HEAD"
I0815 16:16:29.763682  534150 utils.go:461] Found no changed files.
I0815 16:16:29.763749  534150 types.go:72] Loading the catalog for, infra-deployments, from category, tests
I0815 16:16:29.763766  534150 types.go:144] The following rules have matched .
I0815 16:16:29.763772  534150 types.go:169] Will apply rules
```



# Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

## [KONFLUX-3187](https://issues.redhat.com/browse/KONFLUX-3187)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added meaningful description with JIRA/GitHub issue key(if applicable), for example HASSuiteDescribe("STONE-123456789 devfile source") 
- [ ] I have updated labels (if needed)

Dannyb48 commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,170 @@
+	})}}
+
+// CheckNoComponentsFilesChanged checks if in repo infra-deployments changed files other than components
+func CheckNoComponentsFilesChanged(rctx *rulesengine.RuleCtx) bool {
```
There are two different patterns being used for the default rule and the diff files rule. Neither is right or wrong. This comment is more of a consistency of following the same pattern for both rules set. So this one is more of a nitpick. Since you've already written out the rules for each directory, you may want to convert the default rule into a RuleChain as well and us a `None` conditional check around all the rules.

kasemAlem replied:
good point  , 
if I went to implement the default rule based on the RuleChain you mentioned,  
would it it be like the suggested below ? 

```
var InfraDeploymetsDefaultRuleChain = rulesengine.Rule{Name: "Infra Deployments Default Test Execution",
	Description: "Run the Konflux-demo suite tests when an Infra-deployments PR includes changes to files outside of the specified components.",
	Condition:    rulesengine.None{&InfraDeploymentsTestFilesOnlyRule},
	Actions:     []rulesengine.Action{rulesengine.ActionFunc(ExecuteInfraDeploymentsDefaultTestAction)},
}
```

Dannyb48 replied:
@kasemAlem I was thinking something like the below. That is what I have in the example from the e2e-test repo

```
                rulesengine.None{
                        rulesengine.ConditionFunc(CheckPkgFilesChanged),
                        rulesengine.ConditionFunc(CheckMageFilesChanged),
                        rulesengine.ConditionFunc(CheckCmdFilesChanged),
                        rulesengine.ConditionFunc(CheckNoFilesChanged),
                        rulesengine.ConditionFunc(CheckTektonFilesChanged),
                },
```

kasemAlem replied:
got it, 
I've replaced it to use the current rules with **None** , 
also tested

Dannyb48 commented on integration-tests/pipelines/konflux-e2e-tests-pipeline.yaml:
Should these changes be part of this PR?

kasemAlem replied:
was added by mistake, 
removed

Dannyb48 reviewed (commented):
I think the only thing missing to be able to test this is the change in the magefile so that we can test/use against infra-deployments, right? Similar to this one

https://github.com/konflux-ci/e2e-tests/blob/b6d9fa1b514f05df2107e61326687fe2f570e1cb/magefiles/magefile.go#L328

kasemAlem commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,170 @@
+
+// ExecuteInfraDeploymentsDefaultTestAction excutes all the e2e-tests and component suites
+func ExecuteInfraDeploymentsDefaultTestAction(rctx *rulesengine.RuleCtx) error {
+	rctx.LabelFilter = "konflux-demo,release-service,jvm-build-service,image-controller,integration-service,ec,build-templates,multi-platform,build-service"
```
you are right, 
re checked the related [chart](https://miro.com/app/board/uXjVKMPoj9w=/) 
fixed, just set the **konflux-demo** one .

kasemAlem commented:
> I think the only thing missing to be able to test this is the change in the magefile so that we can test/use against infra-deployments, right? Similar to this one
> 
> https://github.com/konflux-ci/e2e-tests/blob/b6d9fa1b514f05df2107e61326687fe2f570e1cb/magefiles/magefile.go#L328

agree, 
I had this part , but for some reason after a rebase it was removed 
I've added it back

psturc commented on magefiles/magefile.go:
```diff
@@ -347,6 +347,28 @@ func RunE2ETests() error {
+		rctx.JobName = jobName
+		rctx.JobType = jobType
+
+		files, err := getChangedFiles()
```
I think this won't work. We need to get list of changed files from within the infra-deployments directory, but this function lists files from the current (e2e-tests) directory.

I'd suggest to update the `getChangedFiles` func to accept the parameter for the location of the directory it should run `git diff` against.
@Dannyb48 wdyt?

kasemAlem replied:
agree, 
it didn't work on the **infra-deployments** , so I've fixed the function 
to get the repo and to check the absolute path of **infra-deployments** 
tested and working fine , see examples in description of PR

psturc commented on magefiles/utils.go:
```diff
@@ -441,8 +441,8 @@ func HandleErrorWithAlert(err error, errLevel slack.ErrorSeverityLevel) error {
-func getChangedFiles() (rulesengine.Files, error) {
-	var gitDiff = sh.OutCmd("git", "diff")
+func getChangedFiles(repo string) (rulesengine.Files, error) {
+	var gitDiff = sh.OutCmd(fmt.Sprintf("git -C ../%s", repo), "diff")
```
Did you had a chance to test this? I think it won't work either, because [infra-deployments is being cloned to `e2e-tests/tmp/` dir](https://github.com/konflux-ci/e2e-tests/blob/e51393f156f418cdb2e5997b1ced6193e369c717/magefiles/installation/install.go#L107), so either we change that or add a condition here for updating the path

kasemAlem replied:
fixed

psturc commented on magefiles/utils.go:
```diff
@@ -441,8 +441,13 @@ func HandleErrorWithAlert(err error, errLevel slack.ErrorSeverityLevel) error {
+	// If the repo is infra-deployments then we check changed files in path tmp/infra-deployments
+	// else it will run locally which is from e2e-tests directory
+	if repo == "infra-deployments" {
+		gitDiff = sh.OutCmd(fmt.Sprintf("git -C tmp/%s", repo), "diff")
```
did you test this? I ran it locally and hit this error
```
exec: git -C tmp/infra-deployments "diff" "--name-status" "main..HEAD"
E0814 16:47:48.715244   82728 utils.go:455] Failed to run git status locally
Error: failed to run "git -C tmp/infra-deployments diff --name-status main..HEAD: fork/exec git -C tmp/infra-deployments: no such file or directory"
```

I think you need to specify the absolute path

kasemAlem replied:
yes, 
ran it locally on both repos  an it worked for me , 
how did you run it ?

psturc replied:
updated the `RunRuleDemo` mage target and ran

```
➜  e2e-tests git:(KONFLUX-3187) ✗ ./mage -v local:runruledemo    
Running target: Local:RunRuleDemo
exec: git -C tmp/infra-deployments "diff" "--name-status" "main..HEAD"
E0814 17:17:55.462549   84437 utils.go:454] Failed to run git status locally
Error: failed to run "git -C tmp/infra-deployments diff --name-status main..HEAD: fork/exec git -C tmp/infra-deployments: no such file or directory"
```

psturc replied:
also when I fixed it, not rules were applied

```
I0814 17:28:47.970449   84868 utils.go:473] The following files, components/build-service/base/kustomization.yaml, were changed!
I0814 17:28:47.970590   84868 types.go:72] Loading the catalog for, infra-deployments, from category, tests
I0814 17:28:47.970595   84868 types.go:144] The following rules have matched .
I0814 17:28:47.970597   84868 types.go:169] Will apply rules
```

Dannyb48 replied:
Hmm, I think what has to be passed into the `getChangedFiles` is the path generated by the installer in order for this to work properly. Especially in CI 

https://github.com/konflux-ci/e2e-tests/blob/59112686eb046d4e8bae4b979fc6e0348150e03c/magefiles/installation/install.go#L107

Not this 

https://github.com/konflux-ci/e2e-tests/pull/1323/files#diff-d7a1577b2d546e34aaf816bdd1b83edef28a66a555f049767bafa83268b7efbbR350

And this if check should just check if the parameter/arg for path is empty. If it is empty, default to the existing way we do the git diff. If it is not, create the git diff using the -c <path arg>. Essentially, lets not force this function to build the path for us. Lets build the path outside of the function

Local dev testing I think that infra-deployments repo should be cloned into that same directory like we do for CI and again we just pass the path to the function. 

Thoughts?

psturc replied:
>Hmm, I think what has to be passed into the getChangedFiles is the path generated by the installer in order for this to work properly

but we'd need to init the installation controller in order to get the path, which I think is an overkill. Other option could be to put that variable to `pkg/consts.go`, so it will be accessible from both install controller and rule engine, wdyt?

> Local dev testing I think that infra-deployments repo should be cloned into that same directory like we do for CI

we actually do clone the infra-deployments to `e2e-tests/tmp/infra-deployments` in CI

Dannyb48 replied:
> I agree but we'd need to init the installation controller in order to get the path, which I think is an overkill.
> 
> Other option could be to put that variable to magefiles/constants/constants.go or pkg/constants/constants.go, so it will be accessible from both install controller and rule engine, wdyt?

I agree, that is what I was thinking as well. That it be some global const that gets used in both context. 

> we actually do clone the infra-deployments to e2e-tests/tmp/infra-deployments in CI

Right, I was just commenting for local evaluating of rules for infra-deployments we would need to have that documented as pre-req. That if you need to clone it into that path

kasemAlem replied:
I've fixed it so now the diff will look on the absolute path 
ran it locally and it works fine . 
WDYT?

psturc commented:
@kasemAlem didn't you get any error when testing it?

```
➜  e2e-tests git:(KONFLUX-3187) ginkgo version  
Ginkgo Version 2.17.3
➜  e2e-tests git:(KONFLUX-3187) ./mage -v local:runruledemo
Running target: Local:RunRuleDemo
exec: git "-C" "/Users/psturc/work/konflux-ci/e2e-tests/tmp/infra-deployments" "diff" "--name-status" "main..HEAD"
I0815 16:03:26.060423   28229 utils.go:474] The following files, components/build-service/README.md, were changed!
I0815 16:03:26.060714   28229 types.go:72] Loading the catalog for, infra-deployments, from category, tests
I0815 16:03:26.060733   28229 types.go:144] The following rules have matched Infra-deployments PR Components File Diff Execution.
I0815 16:03:26.060744   28229 types.go:169] Will apply rules
E0815 16:03:26.061081   28229 common.go:30] failed with [Syntax Error Parsing Label Filter
  ,build-service,konflux-demo
  Operator '||' missing left hand operand.

  Learn more at: http://onsi.github.io/ginkgo/#spec-labels
]
exec: ginkgo "--seed=1723730605" "--timeout=1h30m0s" "--grace-period=30s" "--output-interceptor-mode=none" "--label-filter=,build-service,konflux-demo" "--junit-report=e2e-report.xml" "--p" "--output-dir=." "./cmd" "--"
```
when I ran it, the comma at the beginning of the label filter expression caused a failure ^

psturc commented on magefiles/magefile.go:
```diff
@@ -1382,22 +1398,47 @@ func (Local) PreviewTestSelection() error {
+// 	if err != nil {
+// 		return err
+// 	}
+
+// 	return nil
+// }
+////////////////////
+
 func (Local) RunRuleDemo() error {
 
 	rctx := rulesengine.NewRuleCtx()
-	files, err := getChangedFiles()
-	if err != nil {
-		klog.Error(err)
-		return err
-	}
-	rctx.DiffFiles = files
-	rctx.DryRun = true
+	rctx.Parallel = true
+	rctx.OutputDir = artifactDir
+	// used by ReleaseServiceCatalogRule
+	rctx.IsPaired = fmt.Sprintf("%t", isPRPairingRequired("release-service"))
+	rctx.JUnitReport = "e2e-report.xml"
+	//This conditional could be moved into a rule but keeping the change small
 
-	err = engine.MageEngine.RunRulesOfCategory("demo", rctx)
+	rctx.RepoName = "infra-deployments"
+	rctx.JobName = ""
+	rctx.JobType = ""
 
+	files, err := getChangedFiles("infra-deployments")
 	if err != nil {
 		return err
 	}
+	rctx.DiffFiles = files
 
-	return nil
+	// filtering the rule engine to load only infra-deployments rule catalog within the test category
+	return engine.MageEngine.RunRules(rctx, "tests", "infra-deployments")
 }
```
should this be reverted?

kasemAlem replied:
reverted

psturc commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,158 @@
+
+// ExecuteInfraDeploymentsDefaultTestAction excutes all the e2e-tests and component suites
+func ExecuteInfraDeploymentsDefaultTestAction(rctx *rulesengine.RuleCtx) error {
+	rctx.LabelFilter = "konflux-demo"
```
the label filter name for the demo test is [`konflux`](https://github.com/konflux-ci/e2e-tests/blob/602a159f8ff8b26993209baad53101bd40ace95c/tests/konflux-demo/const.go#L27)

psturc commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,158 @@
+		&InfraDeploymentsJVMTestFileChangeRule},
+	Actions: []rulesengine.Action{rulesengine.ActionFunc(func(rctx *rulesengine.RuleCtx) error {
+		// Adding konflux-demo to the label filtetr when componenets changed
+		rctx.LabelFilter = fmt.Sprintf("%s,%s", rctx.LabelFilter, "konflux-demo")
```
```suggestion
		rctx.LabelFilter = fmt.Sprintf("%s,%s", rctx.LabelFilter, "konflux")
```

psturc commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,158 @@
+	Description: "Map Integration tests files when Integration component files are changed in the infra-deployments PR",
+	Condition: rulesengine.ConditionFunc(func(rctx *rulesengine.RuleCtx) bool {
+
+		return len(rctx.DiffFiles.FilterByDirGlob("components/integration/*")) != 0
```
I think this expression covers just **files** under `components/integration/` dir. We should be covering [all directories and files under it as well](https://github.com/redhat-appstudio/infra-deployments/tree/main/components/integration)

kasemAlem replied:
I made a change into a yaml file in **components/integration/development/kustomization.yaml** , commit 
then ran the demo ,  looks like the change was identified : 

```
exec: git "-C" "/home/kasemalem/integration-project/e2e-tests/tmp/infra-deployments" "diff" "--name-status" "upstream/main"
I0816 11:31:19.981602 1290396 utils.go:473] The following files, components/integration/OWNERS, components/integration/development/kustomization.yaml, were changed!
I0816 11:31:19.981659 1290396 types.go:72] Loading the catalog for, infra-deployments, from category, tests
I0816 11:31:19.981678 1290396 types.go:144] The following rules have matched Infra-deployments PR Components File Diff Execution.
I0816 11:31:19.981684 1290396 types.go:157] DryRun has been enabled will apply them in dry run mode
exec: ginkgo "--seed=1723797078" "--dry-run" "--timeout=1h30m0s" "--grace-period=30s" "--output-interceptor-mode=none" "--label-filter=integration-service,konflux" "--junit-report=e2e-report.xml" "--p" "--output-dir=." "./cmd" "--"
```
 this means that current rule will be working also of changes made recursively, @psturc WDYT?

psturc replied:
I think it was identified only because it contained the change to `components/integration/OWNERS`

psturc replied:
I just tested it with an update to build service and it doesn't work
```
➜  e2e-tests git:(KONFLUX-3187) ./mage -v local:runruledemo
Running target: Local:RunRuleDemo
exec: git "-C" "/Users/psturc/work/konflux-ci/e2e-tests/tmp/infra-deployments" "diff" "--name-status" "upstream/main"
I0816 12:19:24.863695   60516 utils.go:473] The following files, components/build-service/base/kustomization.yaml, were changed!
I0816 12:19:24.863824   60516 types.go:72] Loading the catalog for, infra-deployments, from category, tests
I0816 12:19:24.863846   60516 types.go:144] The following rules have matched Infra Deployments Default Test Execution.
I0816 12:19:24.863858   60516 types.go:169] Will apply rules
exec: ginkgo "--seed=1723803564" "--timeout=1h30m0s" "--grace-period=30s" "--output-interceptor-mode=none" "--label-filter=konflux" "./cmd" "--"
```

kasemAlem commented:
> @kasemAlem didn't you get any error when testing it?
> 
> ```
> ➜  e2e-tests git:(KONFLUX-3187) ginkgo version  
> Ginkgo Version 2.17.3
> ➜  e2e-tests git:(KONFLUX-3187) ./mage -v local:runruledemo
> Running target: Local:RunRuleDemo
> exec: git "-C" "/Users/psturc/work/konflux-ci/e2e-tests/tmp/infra-deployments" "diff" "--name-status" "main..HEAD"
> I0815 16:03:26.060423   28229 utils.go:474] The following files, components/build-service/README.md, were changed!
> I0815 16:03:26.060714   28229 types.go:72] Loading the catalog for, infra-deployments, from category, tests
> I0815 16:03:26.060733   28229 types.go:144] The following rules have matched Infra-deployments PR Components File Diff Execution.
> I0815 16:03:26.060744   28229 types.go:169] Will apply rules
> E0815 16:03:26.061081   28229 common.go:30] failed with [Syntax Error Parsing Label Filter
>   ,build-service,konflux-demo
>   Operator '||' missing left hand operand.
> 
>   Learn more at: http://onsi.github.io/ginkgo/#spec-labels
> ]
> exec: ginkgo "--seed=1723730605" "--timeout=1h30m0s" "--grace-period=30s" "--output-interceptor-mode=none" "--label-filter=,build-service,konflux-demo" "--junit-report=e2e-report.xml" "--p" "--output-dir=." "./cmd" "--"
> ```
> 
> when I ran it, the comma at the beginning of the label filter expression caused a failure ^

fixed : 

`exec: ginkgo "--seed=1723739862" "--timeout=1h30m0s" "--grace-period=30s" "--output-interceptor-mode=none" "--label-filter=integration-service,konflux" "--junit-report=e2e-report.xml" "--p" "--output-dir=." "./cmd" "--"`

`exec: ginkgo "--seed=1723740122" "--timeout=1h30m0s" "--grace-period=30s" "--output-interceptor-mode=none" "--label-filter=konflux" "--junit-report=e2e-report.xml" "--p" "--output-dir=." "./cmd" "--"`

kasemAlem commented:
> @kasemAlem can you please also create a draft PR on infra-deployments side to pair it with this PR and basically validate this change with it?

created the draft PR https://github.com/redhat-appstudio/infra-deployments/pull/4351

psturc commented on magefiles/utils.go:
```diff
@@ -441,8 +441,15 @@ func HandleErrorWithAlert(err error, errLevel slack.ErrorSeverityLevel) error {
+		wd, _ := os.Getwd()
+		gitDiff = sh.OutCmd("git", "-C", fmt.Sprintf("%s/tmp/%s", wd, repo), "diff")
+	}
 	output, err := gitDiff("--name-status", "main..HEAD")
```
so we [hit an issue in infra-deployments PR](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/redhat-appstudio_infra-deployments/4351/pull-ci-redhat-appstudio-infra-deployments-main-appstudio-e2e-tests/1824130672202092544#1:build-log.txt%3A1460-1464) when using this git diff argument


I think this should fix it:
```suggestion
	output, err := gitDiff("--name-status", "upstream/main")
```

psturc commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,194 @@
+		return len(rctx.DiffFiles.FilterByDirGlob("components/integration/*")) != 0
+	}),
+	Actions: []rulesengine.Action{rulesengine.ActionFunc(func(rctx *rulesengine.RuleCtx) error {
+		if !strings.Contains(rctx.LabelFilter, "integration-service") {
+			if rctx.LabelFilter == "" {
+				rctx.LabelFilter = "integration-service"
+			} else {
+				rctx.LabelFilter = fmt.Sprintf("%s,%s", rctx.LabelFilter, "integration-service")
+			}
+		}
```
this would be worth to put into reusable function

kasemAlem replied:
created function to implement this solution

kasemAlem commented:
Hey @psturc , could be that I'm doing it in a wrong way testing the changes of build service only ? 

### so what I did is to create locally in **tmp/infra-deployments** a new branch , then 
performed two changes for two files each in different folder : 

```
➜  infra-deployments git:(infra-test-selection) vim components/build-service/base/kustomization.yaml 
➜  infra-deployments git:(infra-test-selection) ✗ vim components/build-service/README.md 
➜  infra-deployments git:(infra-test-selection) ✗ git status                                          
On branch infra-test-selection
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   components/build-service/README.md
	modified:   components/build-service/base/kustomization.yaml

no changes added to commit (use "git add" and/or "git commit -a")
➜  infra-deployments git:(infra-test-selection) ✗ git add components/build-service/README.md components/build-service/base/kustomization.yaml 
➜  infra-deployments git:(infra-test-selection) ✗ git commit -m "hhx"
[infra-test-selection 2adafb42] hhx
 2 files changed, 2 insertions(+), 1 deletion(-)
```
### after that I ran the  demo mage file : 
```
➜  e2e-tests git:(KONFLUX-3187) ✗ ./mage -v local:runruledemo                                                                                          
Running target: Local:RunRuleDemo
exec: git "-C" "/home/kasemalem/integration-project/e2e-tests/tmp/infra-deployments" "diff" "--name-status" "upstream/main"
I0818 18:39:43.508994 2434013 utils.go:473] The following files, components/build-service/README.md, components/build-service/base/kustomization.yaml, were changed!
I0818 18:39:43.509036 2434013 types.go:72] Loading the catalog for, infra-deployments, from category, tests
I0818 18:39:43.509051 2434013 types.go:144] The following rules have matched Infra-deployments PR Components File Diff Execution.
I0818 18:39:43.509055 2434013 types.go:157] DryRun has been enabled will apply them in dry run mode
exec: ginkgo "--seed=1723995582" "--dry-run" "--timeout=1h30m0s" "--grace-period=30s" "--output-interceptor-mode=none" "--label-filter=build-service,konflux" "--junit-report=e2e-report.xml" "--p" "--output-dir=." "./cmd" "--"
```

tisutisu commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,155 @@
+		return len(rctx.DiffFiles.FilterByDirGlob("components/jvm-build-service/*")) != 0
+	}),
+	Actions: []rulesengine.Action{rulesengine.ActionFunc(func(rctx *rulesengine.RuleCtx) error {
+		AddLabelToLabelFilter(rctx, "jvm-build-service")
```
since [jvm-build-service tests are skipped](https://github.com/konflux-ci/e2e-tests/blob/7eb9fd8575f3c52d76d4d2736f92ff6a5038e3bb/tests/build/jvm-build.go#L39), will it be good to skip the jobs instead [here](https://github.com/openshift/release/blob/4a1405501f3ee3406a34fcbe4b4f9bb8d58f6c5b/ci-operator/config/redhat-appstudio/infra-deployments/redhat-appstudio-infra-deployments-main.yaml#L47) ito avoid provisioning cluster?

psturc replied:
@tisutisu we still run jbs test in ["konflux" suite](https://github.com/konflux-ci/e2e-tests/blob/7eb9fd8575f3c52d76d4d2736f92ff6a5038e3bb/tests/konflux-demo/konflux-demo.go#L384) - should we disable that one as well?

tisutisu replied:
marked [it as pending](https://github.com/konflux-ci/e2e-tests/blob/7eb9fd8575f3c52d76d4d2736f92ff6a5038e3bb/tests/konflux-demo/konflux-demo.go#L386) as well?

tisutisu commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,155 @@
+		return len(rctx.DiffFiles.FilterByDirGlob("components/multi-platform-controller/*")) != 0
+	}),
+	Actions: []rulesengine.Action{rulesengine.ActionFunc(func(rctx *rulesengine.RuleCtx) error {
+		AddLabelToLabelFilter(rctx, "multi-platform")
```
similar as above, shall since multi-platform tests are skipped [here](https://github.com/konflux-ci/e2e-tests/blob/7eb9fd8575f3c52d76d4d2736f92ff6a5038e3bb/tests/build/multi-platform.go#L63), until that is enabled, no point in adding the labels, will it be good to skip running the jobs instead ?

psturc replied:
good point. although I think it's no harm to keep this rule here - it is at least guaranteed that we don't forget to add the rule once MP tests are enabled :)

should we then have a ticket for disabling the CI for changes to JBS/multi-platform component?

tisutisu replied:
for JBS tests skipping we got confirmation from konflux architect .. it will be reworked.. maybe tests needs to be rewritten as well.. they are yet to create stories in the development side .. and for the multi-platform tests being skipped due to [this issue](https://issues.redhat.com//browse/KFLUXBUGS-1179), until that gets fixed, I think we can use this story  for the skipping CI (openshift-ci job which runs e2e-tests)

psturc replied:
addressed in https://github.com/openshift/release/pull/55741, PTAL

psturc commented:
> Hey @psturc , could be that I'm doing it in a wrong way testing the changes of build service only ?
> 
> ### so what I did is to create locally in **tmp/infra-deployments** a new branch , then
> performed two changes for two files each in different folder :
> ...

@kasemAlem  I was updating only `components/build-service/base/kustomization.yaml`, then I saw the issue

kasemAlem commented:
> > Hey @psturc , could be that I'm doing it in a wrong way testing the changes of build service only ?
> > ### so what I did is to create locally in **tmp/infra-deployments** a new branch , then
> > performed two changes for two files each in different folder :
> > ...
> 
> @kasemAlem I was updating only `components/build-service/base/kustomization.yaml`, then I saw the issue

let me try to run same scenario in the infra-deployments PR I've created https://github.com/redhat-appstudio/infra-deployments/pull/4351

psturc commented on magefiles/magefile.go:
```diff
@@ -247,7 +247,7 @@ func (Local) CleanupPrivateRepos() error {
 	quayOrgToken := os.Getenv("DEFAULT_QUAY_ORG_TOKEN")
 	if quayOrgToken == "" {
-		return fmt.Errorf("%s", quayTokenNotFoundError)
+		return fmt.Errorf(quayTokenNotFoundError)
```
checking out these changes and rebasing should resolve this issue.
```
git pull --rebase <remote-name> main
```

Dannyb48 commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,207 @@
+	Description: "Map Integration tests files when Integration component files are changed in the infra-deployments PR",
+	Condition: rulesengine.ConditionFunc(func(rctx *rulesengine.RuleCtx) bool {
+
+		return len(rctx.DiffFiles.FilterByDirGlob("components/integration/*")) != 0 ||
+			len(rctx.DiffFiles.FilterByDirGlob("components/integration/base/*")) != 0 ||
+			len(rctx.DiffFiles.FilterByDirGlob("components/integration/base/external-secrets/*")) != 0 ||
+			len(rctx.DiffFiles.FilterByDirGlob("components/integration/development/*")) != 0 ||
+			len(rctx.DiffFiles.FilterByDirGlob("components/integration/production/base/*")) != 0 ||
+			len(rctx.DiffFiles.FilterByDirGlob("components/integration/production/stone-prod-p01/*")) != 0 ||
+			len(rctx.DiffFiles.FilterByDirGlob("components/integration/production/stone-prod-p02/*")) != 0 ||
+			len(rctx.DiffFiles.FilterByDirGlob("components/integration/rh-certs/*")) != 0 ||
+			len(rctx.DiffFiles.FilterByDirGlob("components/integration/staging/base/*")) != 0 ||
+			len(rctx.DiffFiles.FilterByDirGlob("components/integration/staging/stone-stage-p01/*")) != 0
```
Yeah, I really like this! I've found the current standard implementation of filepath.match very limiting. If we implement this we can also clean up the logic in e2e-repo in a separate PR

kasemAlem replied:
agree , 
I've tried the below implementation and tested it locally ,  on several scenarios 
and seems working fine :

```
var InfraDeploymentsBuildServiceTestFileChangeRule = rulesengine.Rule{Name: "Infra-deployments PR Build service component File Change Rule",
	Description: "Map build service tests files when Build service component files are changed in the infra-deployments PR",
	Condition: rulesengine.ConditionFunc(func(rctx *rulesengine.RuleCtx) bool {
		return len(rctx.DiffFiles.FilterByDirGlob("components/build-service/**/*.yaml")) != 0 ||
			len(rctx.DiffFiles.FilterByDirGlob("components/build-service/**/*.go")) != 0
	}),
	Actions: []rulesengine.Action{rulesengine.ActionFunc(func(rctx *rulesengine.RuleCtx) error {
		AddLabelToLabelFilter(rctx, "build-service")
		return nil
	})}}
```
this one does not require importing [import the doublestart package](https://github.com/bmatcuk/doublestar?tab=readme-ov-file#installation)

but I'm fine with both methods , wdyt?

kasemAlem replied:
> Yeah, I really like this! I've found the current standard implementation of filepath.match very limiting. If we implement this we can also clean up the logic in e2e-repo in a separate PR

Awesome, 
will start implementing with the double star

psturc replied:
I'd be very surprised if this works with the standard library method. Could you try changing some deeply nested file, e.g. `components/build-service/base/build-pipeline-config/build-pipeline-config.yaml` and paste the log here?

kasemAlem replied:
I've made the implementation ,  tested locally several scenarios 
working as expected

kasemAlem replied:
> components/build-service/base/build-pipeline-config/build-pipeline-config.yaml

```
➜  infra-deployments git:(test-infra) vim components/build-service/base/build-pipeline-config/build-pipeline-config.yaml
➜  infra-deployments git:(test-infra) ✗ git status
On branch test-infra
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   components/build-service/base/build-pipeline-config/build-pipeline-config.yaml

no changes added to commit (use "git add" and/or "git commit -a")
➜  infra-deployments git:(test-infra) ✗ git add components/build-service/base/build-pipeline-config/build-pipeline-config.yaml 
➜  infra-deployments git:(test-infra) ✗ git commit -m "dasa"
[test-infra f20e20c0] dasa
 1 file changed, 1 insertion(+)
```

```
➜  e2e-tests git:(KONFLUX-3187) ✗ ./mage -v local:runInfraDeploymentsRuleDemo
Running target: Local:RunInfraDeploymentsRuleDemo
exec: git "-C" "/home/kasemalem/integration-project/e2e-tests/tmp/infra-deployments" "diff" "--name-status" "upstream/main"
I0820 19:43:55.217851   24459 utils.go:473] The following files, components/build-service/base/build-pipeline-config/build-pipeline-config.yaml, were changed!
I0820 19:43:55.217902   24459 types.go:72] Loading the catalog for, infra-deployments, from category, tests
I0820 19:43:55.217918   24459 types.go:144] The following rules have matched Infra-deployments PR Components File Diff Execution.
I0820 19:43:55.217922   24459 types.go:157] DryRun has been enabled will apply them in dry run mode
exec: ginkgo "--seed=1724172235" "--dry-run" "--timeout=1h30m0s" "--grace-period=30s" "--output-interceptor-mode=none" "--label-filter=build-service,konflux" "--p" "--output-dir=." "./cmd" "--"
```

kasemAlem replied:
>  If we implement this we can also clean up the logic in e2e-repo in a separate PR

@Dannyb48 @psturc I've created a task for implementing double star for the e2e-repo [KONFLUX-4115](https://issues.redhat.com/browse/KONFLUX-4115)

psturc commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,152 @@
+	Description: "Map build-templates tests files when Build-templates component files are changed in the infra-deployments PR",
+	Condition: rulesengine.ConditionFunc(func(rctx *rulesengine.RuleCtx) bool {
+
+		return len(rctx.DiffFiles.FilterByDirGlob("components/build-templates/base/**/*")) != 0
```
shouldn't it be `"components/build-templates/**/*"`?

...or when I think about it, it would maybe make more sense to trigger `build-templates` tests when the [pipeline configmap is updated](https://github.com/redhat-appstudio/infra-deployments/tree/main/components/build-service/base/build-pipeline-config) instead

@tisutisu wdyt?

tisutisu replied:
@psturc  yes you are right. build-definitions changes are promoted through [this ConfigMap](https://github.com/redhat-appstudio/infra-deployments/tree/main/components/build-service/base/build-pipeline-config), it would be good to run `build-templates` labelled tests when `build-pipeline-config.yaml` file is updated.

kasemAlem replied:
then for **build templates**  Rule the condition should be like this 
```suggestion
		return len(rctx.DiffFiles.FilterByDirGlob("components/build-service/base/**/*")) != 0
```
if any changes was made in [components/build-templates](https://github.com/redhat-appstudio/infra-deployments/tree/main/components/build-templates) trigger any specific suite or the default konflux is fine?

psturc replied:
So according to what Sushanta mentioned above, I think we should trigger "build-templates" test, when the related config map is updated, i.e.
```
return len(rctx.DiffFiles.FilterByDirGlob("components/build-service/base/build-pipeline-config/*")) != 0
```

And since this would cover the `build-templates` test scenario, I think we don't need another rule for `components/build-templates` and its subdirectories, i.e. the default test suite (`konflux`) should be run.

@tisutisu please correct me if I'm wrong.

tisutisu replied:
@psturc I am okay with this change ` we should trigger "build-templates" test, when the related config map is updated and don't need another rule for components/build-templates and its subdirectories, i.e. the default test suite (konflux) should be run`

kasemAlem replied:
ok, 
based on that , 
 we should  exclude this change from the components/build-service Rule , right ?

tisutisu replied:
components/build-service rule will be there..  but needs to be modified.. 
- when this file `components/build-service/base/build-pipeline-config/build-pipeline-config.yaml` changed, run `build-templates` labeled tests
- when changes in other sub-directories inside `components/build-service` except that `build-pipeline-config.yaml`, we can run `build-service` labeled  tests

psturc reviewed (commented):
just one question for Sushanta, so let's wait for him

otherwise lgtm

/approve

tisutisu commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,152 @@
+	}),
+		rulesengine.ActionFunc(ExecuteTestAction)}}
+
+var InfraDeploymentsIntegrationTestFileChangeRule = rulesengine.Rule{Name: "Infra-deployments PR Integration component File Change Rule",
```
sorry for the late notice.. will it be good change the naming of the rules for `infra-deployment` repo .. for readability improvement.. rename `InfraDeploymentsIntegrationTestFileChangeRule` to something like `InfraDeploymentsIntegrationComponentChangeRule`

kasemAlem replied:
I'm fine with this, 
@psturc wdyt?

psturc replied:
makes sense to me

psturc commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,153 @@
+	Description: "Map build service tests files when Build service component files are changed in the infra-deployments PR",
+	Condition: rulesengine.ConditionFunc(func(rctx *rulesengine.RuleCtx) bool {
+
+		return len(rctx.DiffFiles.FilterByDirGlob("components/build-service/**/*")) != 0 &&
+			len(rctx.DiffFiles.FilterByDirGlob("components/build-service/base/build-pipeline-config/*")) == 0
```
this doesn't cover the case when someone would update 
* `components/build-service/base/build-pipeline-config/*`
* and some other files under `components/build-service/**/*`
 
but I think that's so rare kind of update that we can address it later if we agree on that.

psturc replied:
yes

psturc reviewed (approved):
/lgtm

just last annoying comment from me!
if we agree on addressing that case in some other PR, feel free to `/unhold`

/hold

psturc commented on magefiles/rulesengine/repos/infra_deployments.go:
```diff
@@ -1,17 +1,166 @@
+		return nil
+	})}}
+
+var InfraDeploymentsBuildServiceTemplatesComponentChangeRule = rulesengine.Rule{Name: "Infra-deployments PR Build service component and templates File Change Rule",
+	Description: "Map build service tests files when Build service component and templates files are changed in the infra-deployments PR",
+	Condition: rulesengine.ConditionFunc(func(rctx *rulesengine.RuleCtx) bool {
+
+		return len(rctx.DiffFiles.FilterByDirGlob("components/build-service/**/*")) != 0 &&
+			len(rctx.DiffFiles.FilterByDirGlob("components/build-service/base/build-pipeline-config/*")) != 0
+	}),
+	Actions: []rulesengine.Action{rulesengine.ActionFunc(func(rctx *rulesengine.RuleCtx) error {
+		AddLabelToLabelFilter(rctx, "build-service")
+		return nil
+	})}}
```
I think this is unnecessary rule and will lead to applying `build-service` rule even when only files under `components/build-service/base/build-pipeline-config/*` are changed.

But IMO this is a behavior we can live with for a while until we fix this.

kasemAlem replied:
thanks @psturc , 
I've tested it locally and you are right , 
please find the fix https://github.com/konflux-ci/e2e-tests/pull/1347

psturc reviewed (approved):
/lgtm

psturc commented:
/unhold
